### ISSUE TITLE
Higher leak memory with track_higher_grad = off.

### ISSUE THREAD

**Issue by MarisaKirisame** [facebookresearch/higher#25]
If i get it correctly, with track_higher_grad = off using it to train should be exactly the same as not using it (automatic differentiation-wise), only that imperative update on tensor is replaced with creating new tensor, and moving the reference to the new pointer.
However, I found out that it still use up more and more memory, especially on diffopt.step(loss).
I think the problem is that pytorch cant distinguish between normal tensor operation and weight update, as they are all functional operation on tensor. Hence, when you do functional update additional graph is still created, forcing old weight to stay in memory.
More specifically, maybe you should call detach() before https://github.com/facebookresearch/higher/blob/master/higher/optim.py#L249
?

Below is a gist that reproduce the memory leak.
https://gist.github.com/MarisaKirisame/e4a48617dbe25eee94f08ab9f7c49d99

**Comment by egrefen** [comment]
Hello @MarisaKirisame. Thanks for filing this issue and for producing a repro. I can try and find some time to look into it next week, but since you have a good hypothesis about where the problem lies and how to solve it, would you mind checking if detaching the grads in `all_grads` conditional on `self._track_higher_grads` being `False` resolves the issue you are seeing? If it does, a pull request would be very appreciated as this will free time to deal with other issues (and my own research 😄).

**Comment by MarisaKirisame** [comment]
Hi, I am working on ICML submission, and we found a workaround that doesnt use higher. I might take a look after ICML, when I have some free time though.

**Comment by egrefen** [comment]
That would be greatly appreciated. Sorry I couldn't fix it in time for you
to be able to use it without the workaround. Good luck with your ICML
submission! :)

**Comment by MarisaKirisame** [comment]
Just FYI, I actually try to add a line of detach and it doesnt fix it. Maybe it is already in the wengert list? I dont know.

**Comment by MarisaKirisame** [comment]
urgh, I am unable to produce after I uninstall and install higher. Maybe this is the old memory leak under disguise?

**Comment by egrefen** [comment]
That would be nice! If you can't repro, I'm going to assume that it was the bug from #6 which was resolved in #15. 

I'm seeing some small but steady increase to memory when running your code with `epochs = 100` on cpu, so I'd like to do a bit of profiling when I have time next week to just sanity check before closing the issue.

In the meantime, please let me know if you observe GPU memory leaks elsewhere now you're working off higher v0.1.5.

**Comment by MarisaKirisame** [comment]
I will let you know.

**Comment by MarisaKirisame** [comment]
Is there any update? I need higher again, and, as I am working on gradient checkpointing, not leaking memory is very crucial. It will also help higher's story as gradient checkpointing make big unrolling possible.

**Comment by egrefen** [comment]
@MarisaKirisame sorry, ICML deadlines and then the UK covid lockdown kind of got in the way of progress here, as I'm the only person working on this currently. I'll see what I can do in the coming week or two.

**Comment by egrefen** [comment]
A few questions for you @MarisaKirisame:
1. Are you still experiencing the memory leak?
2. Is the repro example above still valid?
3. Is this on GPU or CPU?

**Comment by egrefen** [comment]
Just to update. I can confirm there's no memory leak across different unrolls with `track_higher_grad=True` and `False`, but that within an unroll the memory is not being released as it should be when `track_higher_grad=False`. I will try to narrow down there this is happening.

**Comment by MarisaKirisame** [comment]
Thanks! I am being very busy so sorry for not replying earlier.

**Comment by egrefen** [comment]
Fixed in #48. Let me know if you have any further issues. Thanks @MarisaKirisame for flagging this.

**Comment by raulpuric** [comment]
Did anyone ever get gradient checkpointing + higher working? This has been a bottleneck for me and when I checked a few months back gradient checkpointing does not work with second order gradients.

**Comment by MarisaKirisame** [comment]
Thx! the bug is fixed.

**Comment by egrefen** [comment]
> Did anyone ever get gradient checkpointing + higher working? This has been a bottleneck for me and when I checked a few months back gradient checkpointing does not work with second order gradients.

@raulpuric Gradiet checking is not supported at present, and we are unlikely to support it in the short term because we would need a custom implementation. The best thing to do is raise an issue about this so it stays on our radar.

In the meantime, have you tried doing your loop onrolling on CPU? If you're doing MAML-style approaches this isn't much help, but otherwise you can do your main training on GPU.

**Comment by MarisaKirisame** [comment]
@raulpuric why dont gradient checkpointing work for second order gradient?

**Comment by egrefen** [comment]
[The docs](https://pytorch.org/docs/stable/checkpoint.html) say "Checkpointing doesn’t work with `torch.autograd.grad()`, but only with `torch.autograd.backward()`." and we use `torch.autograd.grad()` as part of our differentiable optimizers.

**Comment by egrefen** [comment]
@raulpuric @MarisaKirisame TBH I have not tried it to see how it fails. If you want support for it, I suggest trying it in a simple example, and if it doesn't work, [file an issue](https://github.com/facebookresearch/higher/issues/new) with the example you used and what went wrong.

**Comment by raulpuric** [comment]
will try my best to find some time in between Neurips deadlines.

**Comment by MarisaKirisame** [comment]
the problem seems to be just that pytorch's checkpoint suck. however, it is quite easy to roll your own.

**Comment by MarisaKirisame** [comment]
the bigger problem is both first order gradient and higher order gradient need checkpoint ideally, but the above solution only give checkpoint for first order gradient. getting checkpoint for higher order gradient probably need support here.
